Update processor metrics to obsreport

pjanotti opened the pull request:
**Description:** Initial updates to migrate processor metrics to obsreport package, ie.: the new metrics.

Cleaned-up a bit some of the processor metrics and spelled out the rule names for new metrics.

**Link to tracking Issue:** Related to #141 

**Testing:** Added test for the processor common metrics, validated manually that legacy metrics were still working

**Documentation:** Updated obsreport package doc.go

tigrannajaryan commented on processor/memorylimiter/memorylimiter.go:
```diff
@@ -149,12 +163,24 @@ func (ml *memoryLimiter) ConsumeMetricsData(
+		// 	it is necessary to check the pipeline to see if this is directly connected
+		// 	to a receiver (ie.: a receiver is on the call stack). For now it
+		// 	assumes that the pipeline is properly configured and a receiver is on the
+		// 	callstack.
```
Again I am a bit confused what exactly "refused" means.

Also, I would imagine "dropped" is only counted when we return success to our predecessor but decide not to forward the data to the next component. If that is not how "dropped" is defined then please clarify.

pjanotti replied:
Your understanding of "dropped" is correct. Let me see if I can rewrite to make it more clear.

pjanotti commented on obsreport/obsreport_processor.go:
```diff
@@ -0,0 +1,207 @@
+	// of the collector since processors only deal with internal format.
+	mProcessorAcceptedSpans = stats.Int64(
+		processorPrefix+AcceptedSpansKey,
+		"Number of spans successfully pushed into the pipeline.",
```
That is the intention, an improved description seems in order.

pjanotti commented on obsreport/obsreport_processor.go:
```diff
@@ -0,0 +1,207 @@
+
+	DroppedSpansKey = "dropped_spans"
+
+	DroppedMetricPointsKey = "dropped_metric_points"
```
:) Separation of "areas" but good opportunity to add some comments...

pjanotti commented on obsreport/obsreport.go:
```diff
@@ -83,6 +83,13 @@ func Configure(
+	if configType == "" {
+		return componentPrefix
+	}
+	return componentPrefix + configType + nameSep
```
It is a private function but it won't harm adding the check.

pjanotti commented on obsreport/obsreport_processor.go:
```diff
@@ -0,0 +1,207 @@
+		stats.UnitDimensionless)
+	mProcessorRefusedSpans = stats.Int64(
+		processorPrefix+RefusedSpansKey,
+		"Number of spans that could not be pushed into the pipeline.",
```
I like your suggestion, I will use them.

pjanotti commented on obsreport/obsreport_processor.go:
```diff
@@ -0,0 +1,207 @@
+	td consumerdata.TraceData,
+) {
+	if useNew {
+		processorTraceOp(processorCtx, len(td.Spans), 0, 0)
```
Hum I could remove the inlined function...

owais commented:
Generally LGTM. One question: Do we expect components to not record metrics (dropped) when rejecting spans/metrics (returning error)? So each component only records either dropped when it decides to neither forward, not reject data, or `refused` when the next consumer rejects data (returns error), but a component should not record either when it itself rejects data (returns error). It may record it's own component specific metric. May be it is already present but would be nice to have a bullet points written down for people writing new components.

tigrannajaryan commented on obsreport/obsreport.go:
```diff
@@ -84,6 +85,9 @@ func Configure(
 }
 
 func buildComponentPrefix(componentPrefix, configType string) string {
+	if !strings.HasSuffix(componentPrefix, nameSep) {
```
👍

pjanotti commented:
Good questions @owais, these should be published in a doc so authors can follow them, here is the summary:

## Refused and Dropped Metrics
The intention is to distinguish with certainty if data was lost or not. Refused is to indicate that the data was not stored or buffered by the Collector and actual data loss will depend on the client attempting the operation. Dropped metrics indicate with 100% certainty that there was actual data loss on the Collector. Under this guidance, receivers should not report dropped metrics but only refused. For processors, it is a bit nuanced: if there was a receiver on the call stack it should record the data as refused, but if not it should be dropped.

In principle, the distinction between refused and dropped could be done automatically for processors. For instance, it could use the context passed to data consumers to see if a given call stack is directly connected to a receiver or not. 

In the short run this is going not to be done: instead, the code on each processor will be fixed assuming that the processors are configured for their proper usage, ie.: memory_limiter will report "refused" but not "dropped", while queued_retry will report always "dropped".

The goal in the short run is to get new metrics, properly named, so these internal improvements do not delay the release.

bogdandrutu commented on obsreport/obsreport_processor.go:
```diff
@@ -0,0 +1,212 @@
+	}
+}
+
+func calcNumPoints(md consumerdata.MetricsData) int {
```
return int64 since you always convert that?

bogdandrutu commented on obsreport/obsreport_receiver.go:
```diff
@@ -26,15 +26,23 @@ import (
 	AcceptedSpansKey = "accepted_spans"
-	RefusedSpansKey  = "refused_spans"
+	// Key used to identify spans refused (ie.: not ingested) by the Collector.
+	RefusedSpansKey = "refused_spans"
```
It is better to have all the names prefixed with a namespace that ensures uniqueness. "otel.refused_spans"

pjanotti replied:
They are prefixed (by default) with `otel_col` so in Prometheus they look like:
```terminal
# HELP otelcol_receiver_accepted_metric_points Number of metric points successfully pushed into the pipeline.
# TYPE otelcol_receiver_accepted_metric_points counter
otelcol_receiver_accepted_metric_points{receiver="opencensus",transport="grpc"} 72
# HELP otelcol_receiver_accepted_spans Number of spans successfully pushed into the pipeline.
# TYPE otelcol_receiver_accepted_spans counter
otelcol_receiver_accepted_spans{receiver="jaeger",transport="collector_http"} 8749
otelcol_receiver_accepted_spans{receiver="opencensus",transport="grpc"} 305
otelcol_receiver_accepted_spans{receiver="zipkin",transport="http"} 8670
```
The prefix is done at the app level.

bogdandrutu commented:
PS: I have not fully reviewed but the reason I blocked this PR is addressed
